api/task_manager.cc

```diff
@@ -0,0 +1,130 @@
+    res.type = status.task_status.type;
+    res.state = status.task_status.state;
+    res.is_abortable = bool(status.abortable);
+    res.start_time = st;
+    res.end_time = et;
+    res.error = status.task_status.error;
+    res.parent_id = status.parent_id.to_sstring();
+    res.number = status.task_status.number;
+    res.shard = status.task_status.shard;
+    res.keyspace = status.task_status.keyspace;
+    res.table = status.task_status.table;
+    res.entity = status.task_status.entity;
+    res.progress_units = status.task_status.progress_units;
+    res.progress_total = status.task_status.progress.total;
+    res.progress_completed = status.task_status.progress.completed;
+    return res;
+}
+
+future<json::json_return_type> retrieve_status(task_manager::foreign_task_ptr task) {
+    if (task.get() == nullptr) {
+        return make_exception_future<json::json_return_type>(httpd::bad_param_exception("Invalid task id"));
+    }
+    full_task_status s;
+    s.task_status = task->get_status();
+    s.parent_id = task->get_parent_id();
+    s.abortable = task->is_abortable();
+    s.module = task->get_module();
+    return task->get_progress().then([s] (task_manager::task::progress progress) mutable {
+        s.task_status.progress.completed = progress.completed;
+        s.task_status.progress.total = progress.total;
+        return make_ready_future<json::json_return_type>(make_status(s));
+    });
+}
+
+void set_task_manager(http_context& ctx, routes& r) {
+    tm::get_modules.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return make_ready_future<json::json_return_type>(ctx.tm.local().get_modules());
+    });
+
+    tm::get_tasks.set(r, [&ctx] (std::unique_ptr<request> req) {
+        return do_with(req.get(), [&ctx] (auto &req) {
+            return ctx.tm.map_reduce0([&req] (task_manager& tm) {
+                std::vector<tm::task_stats> res{};
+                std::string m = req->param["module"];
+                if (tm.get_tasks_per_module().contains(m)) {
+                    for (auto& [task_id, task] : tm.get_tasks_per_module()[m]) {
+                        if (filter_tasks(task, req->query_parameters)) {
+                            tm::task_stats ts;
+                            ts.task_id = task->get_status().id.to_sstring();
+                            ts.state = task->get_status().state;
+                            res.push_back(ts);
+                        }
+                    }
+                }
+                return res;
+            }, std::vector<tm::task_stats>(), concat<tm::task_stats>);
+        }).then([] (const std::vector<tm::task_stats>& res) {
+            return make_ready_future<json::json_return_type>(res);
+        });
+    });
+
+    tm::get_task_status.set(r, [&ctx] (std::unique_ptr<request> req) {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        return ctx.tm.local().retrieve_task(id).then(retrieve_status);
```

Comment:
Following the convention of 'X_on_all_shards` in replica::database, looking up a task_id globally can be done using a static task_manager function, e.g.
```
future<foreign_ptr<task_ptr>> lookup_task_on_all_shards(sharded<task_manager>& tm, task_id tid) {
    foreign_ptr<task_ptr> res;
    co_await coroutine::parallel_for_each(boost::irange(0u, smp::count), [&tm, tid] (unsigned shard) -> future<> {
        auto local_res = co_await tm.invoke_on(shard, [tid] (const task_manager& local_tm) {
            const auto& all_tasks = local_tm.get_all_tasks();
            if (auto it = all_tasks.find(tid); it != all_tasks.end()) {
                return make_foreign<task_ptr>(it->second);
            }
            return make_foreign<task_ptr>(nullptr);
        }
        if (!res) {
            res = std::move(local_res);
        } else {
            on_internal_error(tmlogger, format("task_id {} found on more than one shard", tid));
        }
    }
    co_return res;
}
```

Reply:
After copying the status aside, we can check here if the task is done, and unregister it.
This can be done using a task method so it can unregister itself on the right shard, and from the right module on it.

Reply:
The rest_api itself should return an error if the task_id wasn't found.
If it was, the rest_api returns success and the task status (that could hold an error).